### PR TITLE
[WIP ]Implementation of np.linalg.lstsq

### PR DESCRIPTION
**Pull request by joaogui1**
Related to #1999 

- [x] Make sure the implementation works when rank = m = n and when rank = m != n

- [x] Write tests

### CODE REVIEW

**Comment by joaogui1** [comment]
Hey @shoyer can you take a look at this?

**Comment by joaogui1** [comment]
Right now the error is in typechecking, int64 vs int32 and the same for floats, can you point me towards a way to fix this please?

File: tests/linalg_test.py [hunk 1]
```diff
@@ -1131,5 +1131,31 @@ def testIssue2131(self, n, dtype):
...
+    _skip_if_unsupported_type(dtype)
+    rng = rng_factory()
+
+    jnp_fun = lambda lhs, rhs: np.linalg.lstsq(lhs, rhs)
+    onp_fun = lambda lhs, rhs: onp.linalg.lstsq(lhs, rhs)
+
+    a = rng(lhs_shape, dtype)
+    b = rng(rhs_shape, dtype)
+      
+    self.assertAllClose(jnp_fun(a, b), onp_fun(a, b), atol=1e-4, check_dtypes=True)
```
Comment (IvanYashchuk) [inline tests/linalg_test.py]: NumPy always returns rank of `a` with `numpy.int32` type. Current JAX implementation returns `rank` with dtype `int32` or `int64` based on dtype of `a`. It makes sense not to check dtype for `rank`.

File: tests/linalg_test.py [hunk 2]
```diff
@@ -1131,5 +1131,31 @@ def testIssue2131(self, n, dtype):
...
+      {"testcase_name":
+       "_lhs={}_rhs={}".format(
+           jtu.format_shape_dtype_string(lhs_shape, dtype),
+           jtu.format_shape_dtype_string(rhs_shape, dtype)),
+       "lhs_shape": lhs_shape,
+       "rhs_shape": rhs_shape, "dtype": dtype, "rng_factory": rng_factory}
+      for lhs_shape, rhs_shape in [
+          ((2, 4), (2, 2)),
+          ((5, 4), (5, 2)),
+          ((4, 4), (4, 1))
```
Comment (IvanYashchuk) [inline tests/linalg_test.py]: It would be good to add the test case with `rhs_shape = (i, )` (it is different from `(i, 1)`).

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]

File: tests/linalg_test.py (hunk 1 above)
Comment (joaogui1) [reply tests/linalg_test.py]: Didn't know that, thanks!

**Comment by joaogui1** [comment]
Ok, so after explicit casting and @IvanYashchuk help, tests are passing @shoyer

**Comment by joaogui1** [comment]
Hey @mattjj Can you check this out please?

**Pull request closed without merging**